File: src/algorithms/graph/shortest-path/bellman-ford/sources/bellman-ford_test.py
import importlib
import sys
import os
import math

sys.path.insert(0, os.path.dirname(__file__))
module = importlib.import_module("bellman-ford")
bellman_ford = module.bellman_ford


def test_computes_shortest_distances_with_positive_weights():
    adj = {
        "A": [("B", 4), ("C", 2)],
        "B": [("D", 5)],
        "C": [("B", 1), ("D", 8)],
        "D": [],
    }
    distances = bellman_ford(adj, "A", ["A", "B", "C", "D"])
    assert distances["A"] == 0
    assert distances["C"] == 2
    assert distances["B"] == 3
    assert distances["D"] == 8


def test_handles_graph_with_negative_edge_weight():
    adj = {
        "A": [("B", 6), ("C", 7)],
        "B": [("D", 5), ("E", -4)],
        "C": [("D", -3)],
        "D": [("B", -2)],
        "E": [("D", 7)],
    }
    distances = bellman_ford(adj, "A", ["A", "B", "C", "D", "E"])
    assert distances["A"] == 0
    assert isinstance(distances["B"], (int, float))
    assert distances["C"] == 7


def test_returns_zero_for_start_node():
    adj = {"X": [("Y", 3)], "Y": []}
    distances = bellman_ford(adj, "X", ["X", "Y"])
    assert distances["X"] == 0


def test_returns_infinity_for_unreachable_nodes():
    adj = {"A": [("B", 1)], "B": [], "C": []}
    distances = bellman_ford(adj, "A", ["A", "B", "C"])
    assert math.isinf(distances["C"])


def test_handles_single_node_graph():
    adj = {"A": []}
    distances = bellman_ford(adj, "A", ["A"])
    assert distances["A"] == 0


def test_handles_linear_chain_with_mixed_weights():
    adj = {"A": [("B", 3)], "B": [("C", -1)], "C": [("D", 4)], "D": []}
    distances = bellman_ford(adj, "A", ["A", "B", "C", "D"])
    assert distances["B"] == 3
    assert distances["C"] == 2
    assert distances["D"] == 6


def test_marks_nodes_reachable_via_negative_cycle_as_negative_infinity():
    adj = {"A": [("B", 1)], "B": [("C", -3)], "C": [("B", 1)], "D": []}
    distances = bellman_ford(adj, "A", ["A", "B", "C", "D"])
    assert math.isinf(distances["B"]) and distances["B"] < 0


if __name__ == "__main__":
    test_computes_shortest_distances_with_positive_weights()
    test_handles_graph_with_negative_edge_weight()
    test_returns_zero_for_start_node()
    test_returns_infinity_for_unreachable_nodes()
    test_handles_single_node_graph()
    test_handles_linear_chain_with_mixed_weights()
    test_marks_nodes_reachable_via_negative_cycle_as_negative_infinity()
    print("All tests passed!")

File: src/algorithms/graph/shortest-path/bellman-ford/BellmanFord_test.java
import java.util.*;

// Compile: javac BellmanFord.java BellmanFord_test.java
// Run:     java -ea BellmanFord_test
public class BellmanFord_test {
    public static void main(String[] args) {
        testComputesShortestDistancesWithPositiveWeights();
        testHandlesNegativeEdgeWeight();
        testReturnsZeroForStartNode();
        testReturnsInfinityForUnreachableNodes();
        testHandlesSingleNodeGraph();
        testHandlesLinearChainWithMixedWeights();
        testMarksNegativeCycleNodesAsNegativeInfinity();
        System.out.println("All tests passed!");
    }

    static Map<String, List<Object[]>> adj(Object[]... entries) {
        Map<String, List<Object[]>> map = new LinkedHashMap<>();
        for (Object[] entry : entries) {
            String node = (String) entry[0];
            List<Object[]> neighbors = new ArrayList<>();
            for (int edgeIdx = 1; edgeIdx < entry.length; edgeIdx++) {
                neighbors.add((Object[]) entry[edgeIdx]);
            }
            map.put(node, neighbors);
        }
        return map;
    }

    static void testComputesShortestDistancesWithPositiveWeights() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 4}, new Object[]{"C", 2}},
            new Object[]{"B", new Object[]{"D", 5}},
            new Object[]{"C", new Object[]{"B", 1}, new Object[]{"D", 8}},
            new Object[]{"D"}
        );
        Map<String, Double> distances = BellmanFord.bellmanFord(
            adjacencyList, "A", Arrays.asList("A", "B", "C", "D"));
        assert distances.get("A") == 0.0;
        assert distances.get("C") == 2.0;
        assert distances.get("B") == 3.0;
        assert distances.get("D") == 8.0;
    }

    static void testHandlesNegativeEdgeWeight() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 6}, new Object[]{"C", 7}},
            new Object[]{"B", new Object[]{"D", 5}, new Object[]{"E", -4}},
            new Object[]{"C", new Object[]{"D", -3}},
            new Object[]{"D", new Object[]{"B", -2}},
            new Object[]{"E", new Object[]{"D", 7}}
        );
        Map<String, Double> distances = BellmanFord.bellmanFord(
            adjacencyList, "A", Arrays.asList("A", "B", "C", "D", "E"));
        assert distances.get("A") == 0.0;
        assert distances.get("C") == 7.0;
    }

    static void testReturnsZeroForStartNode() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"X", new Object[]{"Y", 3}},
            new Object[]{"Y"}
        );
        Map<String, Double> distances = BellmanFord.bellmanFord(
            adjacencyList, "X", Arrays.asList("X", "Y"));
        assert distances.get("X") == 0.0;
    }

    static void testReturnsInfinityForUnreachableNodes() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 1}},
            new Object[]{"B"},
            new Object[]{"C"}
        );
        Map<String, Double> distances = BellmanFord.bellmanFord(
            adjacencyList, "A", Arrays.asList("A", "B", "C"));
        assert distances.get("C") == Double.MAX_VALUE || distances.get("C").isInfinite();
    }

    static void testHandlesSingleNodeGraph() {
        Map<String, List<Object[]>> adjacencyList = adj(new Object[]{"A"});
        Map<String, Double> distances = BellmanFord.bellmanFord(
            adjacencyList, "A", Arrays.asList("A"));
        assert distances.get("A") == 0.0;
    }

    static void testHandlesLinearChainWithMixedWeights() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 3}},
            new Object[]{"B", new Object[]{"C", -1}},
            new Object[]{"C", new Object[]{"D", 4}},
            new Object[]{"D"}
        );
        Map<String, Double> distances = BellmanFord.bellmanFord(
            adjacencyList, "A", Arrays.asList("A", "B", "C", "D"));
        assert distances.get("B") == 3.0;
        assert distances.get("C") == 2.0;
        assert distances.get("D") == 6.0;
    }

    static void testMarksNegativeCycleNodesAsNegativeInfinity() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 1}},
            new Object[]{"B", new Object[]{"C", -3}},
            new Object[]{"C", new Object[]{"B", 1}},
            new Object[]{"D"}
        );
        Map<String, Double> distances = BellmanFord.bellmanFord(
            adjacencyList, "A", Arrays.asList("A", "B", "C", "D"));
        assert distances.get("B") == Double.NEGATIVE_INFINITY;
    }
}
